clustering/interface.py

```diff
@@ -0,0 +1,96 @@
+from clustering_settings import MAX_RESULTS_FOR_CLUSTERING, CLUSTERING_CACHE_TIME, DEFAULT_FEATURES
+from . import CLUSTERING_RESULT_STATUS_PENDING, CLUSTERING_RESULT_STATUS_FAILED
+
+
+def get_sound_ids_from_solr_query(query_params):
+    """Performs Solr query and returns results as a list of sound ids.
+
+    This method performs a single query to Solr with a very big page size argument so all results are 
+    returned at once. A very big page size will make the clustering take a lot of time to be performed.
+    The number of results to retrieve is defined in the clustering settings file as MAX_RESULTS_FOR_CLUSTERING.
+
+    Args:
+        query_params (dict): contains the query parameters to replicate the user query.
+    
+    Returns
+        List[int]: list containing the ids of the retrieved sounds.
+    """
+    current_page = 1
+
+    # We set include_facets to False in order to reduce the amount of data that Solr will return.
+    query_params.update({
+        'current_page': 1,
+        'sounds_per_page': MAX_RESULTS_FOR_CLUSTERING,
+        'include_facets': False,
+    })
+    query = search_prepare_query(**query_params)
+    _, _, _, _, docs = perform_solr_query(query, current_page)
+
+    resultids = [d.get("id") for d in docs]
+    return resultids
+
+
+def cluster_sound_results(request, features=DEFAULT_FEATURES):
+    """Performs clustering on the search results of the given search request with the requested features.
+
+    This is the main entry to the clustering method. It will either get the clustering results from cache, 
+    or compute it (and store it in cache). When needed, the clustering will be performed async by a celery 
+    worker. 
+
+    Args:
+        request (HttpRequest): request associated with the search query submited by the user.
+        features (str): name of the features to be used for clustering. The available features are defined in the 
+            clustering settings file.
+
+    Returns:
+        Dict: contains either the state of the clustering ('pending' or 'failed') or the resulting clustering classes 
+            and the graph in node-link format suitable for JSON serialization.
+    """
+    query_params, _, extra_vars = search_prepare_parameters(request)
+    # We change filter_query to filter_query_non_facets in order to ensure that the clustering is always
+    # done on the non faceted filtered results. Without that, people directly requesting a facet filtered
+    # page would have a clustering performed on filtered results.
+    query_params['filter_query'] = extra_vars['filter_query_non_facets']
+
+    cache_key = 'cluster-results-{search_query}-{filter_query}-{sort}-{tag_weight}-{username_weight}-{id_weight}-' \
+                '{description_weight}-{pack_tokenized_weight}-{original_filename_weight}-{grouping}'.format(**query_params) \
+                .replace(' ', '')
+
+    cache_key += '-{}'.format(features)
+
+    cache_key_hashed = hash_cache_key(cache_key)
```

Comment:
I can't remember if we discussed this or not. I guess we do this because the memcache key size is quite low? In my experience, hashing cache keys is really annoying, because you can't inspect them easily. 
I guess this is all we can do for now, but I think that perhaps we should move to redis for the cache some time so that we can have large keys and remove the hashing